libdnf/module/module_query.cpp

```diff
@@ -100,4 +100,44 @@ void ModuleQuery::filter_arch(const std::vector<std::string> & patterns, libdnf:
 }
 
 
+bool ModuleQuery::latest_cmp(const ModuleItem * module_item_1, const ModuleItem * module_item_2) {
+    Pool * pool = module_item_1->module_sack->p_impl->pool;
+    const Solvable * s1 = pool_id2solvable(pool, module_item_1->id.id);
+    const Solvable * s2 = pool_id2solvable(pool, module_item_2->id.id);
+    if (s1->name != s2->name) {
+        return s1->name < s2->name;
+    }
+    if (s1->arch != s2->arch) {
+        return s1->arch < s2->arch;
+    }
+    return module_item_1->get_version() > module_item_2->get_version();
+}
+
+
+void ModuleQuery::filter_latest(int limit) {
+    std::vector<const ModuleItem *> same_nsca_vector;
+    same_nsca_vector.reserve(size());
+
+    for (auto & module_item : get_data()) {
+        same_nsca_vector.push_back(&module_item);
+    }
+    sort(same_nsca_vector.begin(), same_nsca_vector.end(), latest_cmp);
+
+    std::string last_nsca = same_nsca_vector.front()->get_name_stream_staticcontext_arch();
+    int kept_in_query = 0;
+    for (auto module_item : same_nsca_vector) {
+        std::string nsca = module_item->get_name_stream_staticcontext_arch();
+        if (last_nsca != nsca) {
+            last_nsca = nsca;
+            kept_in_query = 0;
+        }
+        if (kept_in_query < limit) {
+            kept_in_query += 1;
+        } else {
+            get_data().erase(*module_item);
+        }
+    }
```

Comment:
There is a difference in logic in comparison to PackageQuery. If I have set of packages - dnf-1-1.noarch, dnf-2-2.noarch, dnf-2-2.noarch - PackageQuery.filter_latest(2) will return all 3packages, because it returns 2 versions not 2 latest object. Your code with nodejs:master:asssd:1:noarch, nodejs:master:asssd:2:noarch, nodejs:master:asssd:2:noarch, - ModuleQuery.filter_latest(2) will return 2 module items with version 2. I would recommend to use the same logic like PackageQuery.
```
    std::string last_nsca = same_nsca_vector.front()->get_name_stream_staticcontext_arch();
    long long latest version = -1; // invalid value for version to ensure that it cannot be in module item
    int kept_in_query = 0;
    for (auto module_item : same_nsca_vector) {
        std::string nsca = module_item->get_name_stream_staticcontext_arch();
        if (last_nsca != nsca) {
            last_nsca = nsca;
            kept_in_query = 0;
        }
        if (latest version == module_item->get_version()) {  // do nothing => keep it in query, do not increase kept_in_query
        } else if (kept_in_query < limit) {
            kept_in_query += 1;
        } else {
            get_data().erase(*module_item);
        }
```
What do you think?

Reply:
I changed it to be the same as in PackageQuery. I had to also change the tests a bit (added a new repo with duplicit modulemd documents).